src/man/man1/mbtime.1
.TH mbtime 1 "3 June 2013" "MB-System 5.0" "MB-System 5.0"
.SH NAME
\fBmbtime\fP \- Translate between calendar time values and unix time.

.SH VERSION
Version 5.0

.SH SYNOPSIS
.nf
\fBmbtime\fP
[--calendar-time=year/month/day/hour/minute/second {-Tyear/month/day/hour/minute/second}
--epoch-time=time_d {-Mtime_d}
--help {-H}
--verbose {-V}]
.fi

.SH DESCRIPTION
The program \fBmbtime\fP converts time values between unix time (seconds since
1970/01/01 00:00:00.000000) and calendar time (e.g. 2008/006/05/17/24/32.000000).
Unix time is also known as POSIX time or epoch time, and are typically represented
in \fBMB-System\fP documentation (and source code) as "time_d" values.
The input time is set using the command line arguments \fB\-M\fP\fItime_d\fP for
epoch seconds and \fB\-T\fP\fIyear/month/day/hour/minute/second\fP for
calendar time. The output time (in the form not specified as input) is
written to stdout.

.SH MB-SYSTEM AUTHORSHIP
David W. Caress
.br
  Monterey Bay Aquarium Research Institute
.br
Dale N. Chayes
.br
  Center for Coastal and Ocean Mapping
.br
  University of New Hampshire
.br
Christian do Santos Ferreira
.br
  MARUM - Center for Marine Environmental Sciences
.br
  University of Bremen

.SH OPTIONS
.TP
.B \-\-help \-H
This "help" flag cause the program to print out a description
of its operation and then exit immediately.
.TP
.B \-\-epoch\-time=time_d \-Mtime_d
Sets the input time to be \fItime_d\fP in unix time, which
is defined as seconds since 00:00:00.000000 01 January 1970. If \fB\-M\fP is
specified, this time will be converted to calendar time and output
to stdout in the form
 	yyyy/mm/dd/hh/mm/ss.ssssss
.TP
.B \-\-calendar\-time=year/month/day/hour/minute/second \-Tyear/month/day/hour/minute/second
Sets the input time to be \fIyear/month/day/hour/minute/second\fP.
If \fB\-T\fP is specified, this time will be converted to
\fItime_d\fP in epoch (Unix) seconds, which
are seconds since 00:00:00.000000 01 January 1970.
The \fItime_d\fP value will be output to stdout.
.TP
.B \-\-verbose \-V
Normally, \fBmbtime\fP only prints out the converted time.  If the
\fB\-V\fP flag is given, then \fBmbtime\fP works in a "verbose" mode and
also outputs the program version being used.

.SH EXAMPLES
Suppose one wishes to convert the current time (06 June 2008 18:37:14.000000)
to unix seconds. The following command:
 	mbtime \-T2008/06/06/18/37/14.0
.br
yields the following result:
 	1212777434.000000

Suppose on wishes to convert the unix time 1212777434.0 to calendar time.
The following command:
 	mbtime \-M1212777434.0
.br
yields the following result:
 	2008/06/06/18/37/14.000000

.SH SEE ALSO
\fBmbsystem\fP(1), \fBmbio\fP(1)

.SH BUGS
Never bugs, but perhaps our good friends, the terrestrial isopods.


NEWS
-------------------------------------------------------------------------------
MB-SYSTEM NEWS FILE:
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------

In the GNU package standard, the NEWS file lists the history of user-visible
changes to a package. For the the MB-System open source software package for the
processing and display of swath sonar data, this information is included in the
ChangeLog file. Both the NEWS file and the ChangeLog file are located at the top
of the MB-System source code distribution directory structure.

-------------------------------------------------------------------------------


src/utilities/mbtime.cc
/*--------------------------------------------------------------------
 *    The MB-system:	mbtime.c	6/5/2008
 *
 *    Copyright (c) 2008-2025 by
 *    David W. Caress (caress@mbari.org)
 *      Monterey Bay Aquarium Research Institute
 *      Moss Landing, California, USA
 *    Dale N. Chayes 
 *      Center for Coastal and Ocean Mapping
 *      University of New Hampshire
 *      Durham, New Hampshire, USA
 *    Christian dos Santos Ferreira
 *      MARUM
 *      University of Bremen
 *      Bremen Germany
 *     
 *    MB-System was created by Caress and Chayes in 1992 at the
 *      Lamont-Doherty Earth Observatory
 *      Columbia University
 *      Palisades, NY 10964
 *
 *    See README.md file for copying and redistribution conditions.
 *--------------------------------------------------------------------*/
/*
 * MBTIME converts time values between epoch seconds (seconds since
 * 1970/01/01 00:00:00.000000) and calendar time (e.g. 2008/006/05/17/24/32/0).
 * The input time is set using the command line arguments -Mtime_d for
 * epoch seconds and -Tyear/month/day/hour/minute/second/microsecond for
 * calendar time. The output time (in the form not specified as input) is
 * written to stdout.
 *
 * Author:	D. W. Caress
 * Date:	June 5, 2008
 */

#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#ifdef _WIN32
#include "unistd_w.h"
#else
#include <unistd.h>
#endif
#include "mb_define.h"
#include "mb_status.h"

typedef enum {
    MBTIME_INPUT_EPOCH = 0,
    MBTIME_INPUT_CALENDAR  = 1
} time_mode_t;

constexpr char program_name[] = "MBTIME";
constexpr char help_message[] =
    "MBTIME converts time values between epoch seconds (seconds since\n"
    "1970/01/01 00:00:00.000000) and calendar time (e.g. 2008/006/05/17/24/32/0).\n"
    "The input time is set using the command line arguments -Mtime_d for\n"
    "epoch seconds and -Tyear/month/day/hour/minute/second/microsecond for\n"
    "calendar time. The output time (in the form not specified as input) is\n"
    "written to stdout.";
constexpr char usage_message[] =
    "mbtime [-Mtime_d -Tyear/month/day/hour/minute/second -V -H]\n"
    "\t--calendar-time=year/month/day/hour/minute/second {-Tyear/month/day/hour/minute/second}\n"
    "\t--epoch-time=time_d {-Mtime_d}\n"
    "\t--help {-H}\n"
    "\t--verbose {-V}\n";

/*--------------------------------------------------------------------*/

int main(int argc, char **argv) {
	int verbose = 0;
	time_mode_t mode = MBTIME_INPUT_EPOCH;
	int time_i[7];
	double time_d = 0.0;
	double seconds;

	/* process argument list */
	{
		static struct option options[] = {{"verbose", no_argument, nullptr, 0},
		                                  {"help", no_argument, nullptr, 0},
		                                  {"epoch-time", required_argument, nullptr, 0},
		                                  {"calendar-time", required_argument, nullptr, 0},
		                                  {nullptr, 0, nullptr, 0}};

		int option_index;
		bool errflg = false;
		int c;
		bool help = false;
		while ((c = getopt_long(argc, argv, "VvHhM:m:T:t:", options, &option_index)) != -1)
			switch (c) {
			/* long options all return c=0 */
			case 0:
				if (strcmp("verbose", options[option_index].name) == 0) {
					verbose++;
				}
				else if (strcmp("help", options[option_index].name) == 0) {
					help = true;
				}
				else if (strcmp("epoch-time", options[option_index].name) == 0) {
					sscanf(optarg, "%lf", &time_d);
					mode = MBTIME_INPUT_EPOCH;
				}
				else if (strcmp("calendar-time", options[option_index].name) == 0) {
					sscanf(optarg, "%d/%d/%d/%d/%d/%lf", &time_i[0], &time_i[1], &time_i[2], &time_i[3], &time_i[4], &seconds);
					time_i[5] = (int)seconds;
					time_i[6] = (int)(1000000 * (seconds - time_i[5]));
					mode = MBTIME_INPUT_CALENDAR;
				}
				break;
			case 'H':
			case 'h':
				help = true;
				break;
			case 'M':
			case 'm':
				sscanf(optarg, "%lf", &time_d);
				mode = MBTIME_INPUT_EPOCH;
				break;
			case 'T':
			case 't':
				sscanf(optarg, "%d/%d/%d/%d/%d/%lf", &time_i[0], &time_i[1], &time_i[2], &time_i[3], &time_i[4], &seconds);
				time_i[5] = (int)seconds;
				time_i[6] = (int)(1000000 * (seconds - time_i[5]));
				mode = MBTIME_INPUT_CALENDAR;
				break;
			case 'V':
			case 'v':
				verbose++;
				break;
			case '?':
				errflg = true;
			}

		if (errflg) {
			fprintf(stderr, "usage: %s\n", usage_message);
			fprintf(stderr, "\nProgram <%s> Terminated\n", program_name);
			exit(MB_ERROR_BAD_USAGE);
		}

		if (verbose == 1 || help) {
			fprintf(stdout, "\nProgram %s\n", program_name);
			fprintf(stdout, "MB-system Version %s\n", MB_VERSION);
		}

		if (verbose >= 2) {
			fprintf(stderr, "\ndbg2  Program <%s>\n", program_name);
			fprintf(stderr, "dbg2  MB-system Version %s\n", MB_VERSION);
			fprintf(stderr, "dbg2  Control Parameters:\n");
			fprintf(stderr, "dbg2       verbose:    %d\n", verbose);
			fprintf(stderr, "dbg2       help:       %d\n", help);
			fprintf(stderr, "dbg2       mode:       %d\n", mode);
			fprintf(stderr, "dbg2       time_i[0]:  %d\n", time_i[0]);
			fprintf(stderr, "dbg2       time_i[1]:  %d\n", time_i[1]);
			fprintf(stderr, "dbg2       time_i[2]:  %d\n", time_i[2]);
			fprintf(stderr, "dbg2       time_i[3]:  %d\n", time_i[3]);
			fprintf(stderr, "dbg2       time_i[4]:  %d\n", time_i[4]);
			fprintf(stderr, "dbg2       time_i[5]:  %d\n", time_i[5]);
			fprintf(stderr, "dbg2       time_i[6]:  %d\n", time_i[6]);
			fprintf(stderr, "dbg2       time_d:     %f\n", time_d);
		}

		if (help) {
			fprintf(stderr, "\n%s\n", help_message);
			fprintf(stderr, "\nusage: %s\n", usage_message);
			exit(MB_ERROR_NO_ERROR);
		}

	}

	/* convert to calendar time and output */
	if (mode == MBTIME_INPUT_EPOCH) {
		mb_get_date(verbose, time_d, time_i);
		fprintf(stdout, "%4.4d/%2.2d/%2.2d/%2.2d/%2.2d/%2.2d.%6.6d\n",
			time_i[0], time_i[1], time_i[2], time_i[3], time_i[4],
		        time_i[5], time_i[6]);
	} else {
		/* convert to epoch time and output */
		mb_get_time(verbose, time_i, &time_d);
		fprintf(stdout, "%f\n", time_d);
	}

	if (verbose >= 2) {
		fprintf(stderr, "\ndbg2  Program <%s> completed\n", program_name);
		fprintf(stderr, "dbg2  Ending status:\n");
		fprintf(stderr, "dbg2       status:  %d\n", MB_SUCCESS);
	}

	exit(MB_ERROR_NO_ERROR);
}
/*--------------------------------------------------------------------*/
